kristoffSC/connectors
# [Draft] PR 10 - get table schema from Delta Log.

PR Plan:
PR 5 - Option support in Bounded Mode
PR 6 - Continuous mode without monitoring for changes
PR 6.1 - Package Refactoring
PR 7 - Monitoring for changes in Continuous mode
PR 7.1 - More tests for PR 7 including IT case tests for Continuous mode.
PR 9 - Builder
**PR 10 - Builder refactoring to extract Schema from Delta Log.**
PR 11 - Partition Support using Delta log to identify partition columns.
PR 12 - Get BATCH_SIZE for Format builder from Source options - this value is currently hardcoded. And additional  builder option validation and type safety.

-- All functionalities are in place up to this point. The following PRs are extra tests, eventual bug fixes and code/javadoc adjustments. --

~~PR 11 - Annotate with @deprecated  Sink's builder methods that names starts from "with". Add new ones without "with" prefix and overload partitions with List argument for Sink.~~

PR 11 - Add validation for option names and values including using not applicable options for given mode via generic option method.
PR 12 - Additional IT tests including all options and end to end tests using Source and Sink.

While @kristoffSC is on vacation
@tdas + @scottsand-db to review public API
Once Krzysztof is back
README / docs
QA
end-to-end integration tests - using Sink to write, Source to read, etc.
more examples / migrate examples from internal flink module to examples folder

TODOs and other improvements planed as next:

- Use index for keeping track of already processed paths in SnapshotProcessor - needs new Delta API.
- Do do not create snapshot when streaming only changes (startingVersion or startingTimestamp option used). - needs new Delta API since for startingTimestamp we still need to create a snapshot. There is no API like deltaLog.getChangesFromTimestamp(....)
- Refactor Test util classes to make them common for sink and source.
- Re-visit Delta Standalone APIs for when the source see's metadata/protocol actions. Update ActionProcessor class to handle MetaData and Protocol Actions
- Enhanced DeltaLog.java with `Iterator<VersionLog> getChangesFromTimestamp(long timestamp, boolean failOnDataLoss);` and us it in Delta Source connector for better support of `startingTimestamp` option.
